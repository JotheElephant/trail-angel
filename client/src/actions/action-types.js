const actionTypes = {
  TEST_INCREMENT: 'TEST_INCREMENT',
  TEST_DECREMENT: 'TEST_DECREMENT',
  
  CREATE_USER: 'CREATE_USER',
  UPDATE_USER: 'UPDATE_USER',
  DELETE_USER: 'DELETE_USER',
  RECEIVE_USER_DATA: 'RECEIVE_USER_DATA',

  REQUEST_TRAILS: 'REQUEST_TRAILS',
  FETCH_TRAILS: 'FETCH_TRAILS',
  RECEIVE_TRAILS: 'RECEIVE_TRAILS',
  UPDATE_TRAIL: 'UPDATE_TRAIL',

  SUBMIT_SEARCH: 'SUBMIT_SEARCH',
  RECEIVE_SEARCH_RESULTS: 'RECEIVE_SEARCH_RESULTS',

  REQUEST_FAVORITES: 'REQUEST_FAVORITES',
  FETCH_FAVORITES: 'FETCH_FAVORITES',
  RECEIVE_FAVORITES: 'RECEIVE_FAVORITES',
  ADD_FAVORITE: 'ADD_FAVORITE',
  REMOVE_FAVORITE: 'REMOVE_FAVORITE',

  INITIALIZE_APP: 'INITIALIZE_APP',
  GET_GEOLOCATION: 'GET_GEOLOCATION',
  RECEIVE_GEOLOCATION: 'RECEIVE_GEOLOCATION'
}

export default actionTypes;