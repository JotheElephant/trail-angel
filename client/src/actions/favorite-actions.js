import actionTypes from './action-types';
import dataApi from '../api';

const requestFavorites = (userId) => {
  return {
    type: actionTypes.REQUEST_FAVORITES,
    userId
  };
};

const receiveFavorites = (favorites) => {
  return {
    type: actionTypes.RECEIVE_FAVORITES,
    favorites
  };
};

export const fetchFavorites = (userId) => {
  return (dispatch) => {
    dispatch(requestFavorites(userId));

    // get favorites for user from the server
    return dataApi.trailAngelApi.getFavorites(userId)
      .then((data) => {
        if (data !== undefined && Array.isArray(data)) {
          let promises = data.map((item) => {
            // todo: check to see if the favorite is already in trailsReducer.trails
            // todo: can this just use existing functionality in trail-actions.js?
            // todo: for now, just pull directly from yelp
            return dataApi.yelp({ id: item });
          });

          return Promise.all(promises)
            .then((results) => {
              return dispatch(receiveFavorites(results));
            });
        } // else Promise.reject()?
      });
  };
};

export const addFavorite = (userId, itemId) => {
  return (dispatch) => {
    return data.trailAngelApi.addFavorite(userId, itemId)
      .then(() => {
        dispatch({
          type: actions.ADD_FAVORITE,
          userId,
          itemId
        });
      });
  };
};

export const removeFavorite = (userId, itemId) => {
  return {
    type: actions.REMOVE_FAVORITE,
    userId,
    itemId
  };
};