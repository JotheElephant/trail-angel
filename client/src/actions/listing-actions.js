'use strict';

import * as _ from 'lodash';
import { AsyncStorage } from 'react-native';

import actionTypes from './action-types';
import dataApi from '../api';

const listingActions = {
  getListings: (options = {}) => {
    return (dispatch, getState) => {
      const search = JSON.stringify(options);
      const searches = getState().listingsReducer.searches;
      const cache = getState().listingsReducer.cache;

      return new Promise((resolve, reject) => {

        // if we don't need to search because IDs were passed in with `options` parameter
        if (options.id !== undefined || options.IDs !== undefined) {
          // todo load favorites call is also bringing in empty IDs property for options, going straight to catch block
          return getListingById(options.id || options.IDs, cache)
            .then((results) => {
              dispatch(storeResults(search, results, options.collection));
              return results;
            })
            .catch((err) => {
              console.error('Error getting listings by ID: ', err);
            });
        } else {
          // otherwise search based on search terms
          const cachedListing = findInCache(search, searches, cache);

          // if the search isn't cached, fetch the data
          if (!cachedListing) {

            dispatch(fetchListings());

            dataApi.yelp(options)
              .then((results) => {
                if (results === undefined) {
                  return resolve(false);
                }

                dispatch(storeResults(search, results, options.collection))
                  .then((results) => {
                    resolve(results);
                  });
              })
              .catch((err) => {
                console.error('Error retrieving listing data', err);
                reject(err);
              });
          } else {
            // otherwise, pull from local storage and send no search to save
            dispatch(receiveListings(cachedListing, false));
            resolve(true);    // ???
          }
        }
      });
    };
  },

  cancelRequest: () => {
    return (dispatch) => {
      dispatch({
        type: actionTypes.CANCEL_REQUEST
      });

      dispatch(receiveListings([]));
    }
  },

  updateListings: (listings) => {
    return (dispatch, getState) => {
      return new Promise((resolve, reject) => {
        const updatedListings = {};
        const storageTuples = [];
        const cache = getState().listingsReducer.cache;

        listings.forEach((listing) => {
          if (cache[listing.id] !== undefined) {
            const cachedObject = cache[listing.id];
            updatedListings[listing.id] = {
              ...cachedObject,
              ...listing
            };

            storageTuples.push([ listing.id,  JSON.stringify(listing)]);
          }
        });

        AsyncStorage.multiMerge(storageTuples)
          .then(() => {
            dispatch({
              type: actionTypes.UPDATE_LISTINGS,
              updatedListings
            });
            return resolve();
          })
          .catch((err) => {
            console.error('Error updating listings: ', err);
            return reject(err);
          });
      });
    };
  },

  loadFavorites: () => {
    return (dispatch, getState) => {
      return new Promise((resolve, reject) => {
        const userId = getState().userReducer.userId;
        dataApi.trailAngelApi.getFavorites(userId)
          .then((IDs) => {
            if (IDs !== undefined && Array.isArray(IDs)) {
              dispatch(listingActions.getListings({IDs, collection: 'favorites'}))
                .then((listings) => {
                  // ensure that every favorite from the server is marked as a favorite locally
                  listings.forEach((listing) => {
                    dispatch(listingActions.addToCollection(listing.id, 'favorites'));
                  });
                  resolve(listings);
                });
            } else {
              reject(new Error('Error loading favorites: invalid input'));
            }
          })
          .catch((err) => {
            console.error('Error loading favorites: ', err);
            reject(err);
          });
      });
    };
  },

  addToCollection: (id, collectionName) => {
    return (dispatch, getState) => {
      const cache = getState().listingsReducer.cache;
      const collections = getState().listingsReducer.collections;
      const item = cache[id];

      const collectionArray = item.collections === undefined ? [] : item.collections;

      dispatch(listingActions.updateListings([
        {
          ...item,
          collections: [
            ...collectionArray,
            collectionName
          ]
        }
      ]));

      if (_.find(collections[collectionName], {'id': id}) === undefined) {
        const modifiedCollection = [...collections[collectionName], cache[id]];

        dispatch(updateCollection(collectionName, modifiedCollection));

        AsyncStorage.setItem(`collection:${collectionName}`, JSON.stringify(modifiedCollection));
      }
    };
  },

  removeFromCollection: (id, collectionName) => {
    return (dispatch, getState) => {
      const cache = getState().listingsReducer.cache;
      const collections = getState().listingsReducer.collections;
      const item = {...cache[id]};
      if (item.collections === undefined) {
        return;
      }
      const i = item.collections.indexOf(collectionName);

      if (i !== -1) {
        item.collections.splice(i, 1);
      }

      dispatch(listingActions.updateListings([
        {
          ...cache[id],
          collections: item.collections
        }
      ]));

      const j = _.find(collections[collectionName], {'id': id});

      if (j !== undefined) {
        const modifiedCollection = [...collections[collectionName]];
        modifiedCollection.splice(j, 1);

        dispatch(updateCollection(collectionName, modifiedCollection));

        AsyncStorage.setItem(`collection:${collectionName}`, JSON.stringify(modifiedCollection));
      }
    };
  }
};

const getListingById = (IDs, cache) => {
  if (!Array.isArray(IDs)) IDs = [IDs];
  let promises = IDs.map((id) => {
    const cachedListing = _.find(cache, { 'id': id });
    if (cachedListing === undefined) {
      return dataApi.yelp({ id: id })
        .then((data) => {
          return data;
        })
        .catch((err) => {
          console.error('Error fetching listing by ID: ', err);
        });
    } else {
      return cachedListing;
    }
  });

  return Promise.all(promises);
};

const findInCache = (search, searches, cache) => {
  const cached = searches[search];
  if (cached === undefined) return false;

  const listingIDs = cached.results;
  return _.map(listingIDs, (id) => {
    const listing = _.find(cache, { 'id': id });
    if (listing !== undefined) {
      return listing;
    }
  });
};

const storeResults = (search, results, collection) => {
  return (dispatch) => {
    return new Promise((resolve, reject) => {
      // saved search includes an array of result IDs for later lookup
      const IDs = _.map(results, (data) => {
        return data.place_id;
      });

      const searchToSave = {
        type: 'search',
        search,
        results: IDs
      };

      // create the array of tuples that AsyncStorage uses
      const listings = results.map((result) => {
        return [result.id, JSON.stringify({...result, type: 'listing' })];
      });

      // update the in-memory store with the data and search/result object
      dispatch(receiveListings(results, {  search,
        results: _.map(results, 'id')}, collection));

      // store the saved search with results IDs
      return AsyncStorage.setItem(search, JSON.stringify(searchToSave))
        .then(() => {
          // store the raw data for listingsReducer's cache object
          return AsyncStorage.multiSet(listings);
        });
    });
  };
}

const fetchListings = () => {
  return {
    type: actionTypes.FETCH_LISTINGS
  };
};

const receiveListings = (searchResults, searchToSave, collection) => {
  return {
    type: actionTypes.RECEIVE_LISTINGS,
    searchResults,
    searchToSave,
    collection
  }
};

const updateCollection = (name, collection) => {
  return {
    type: actionTypes.UPDATE_COLLECTION,
    name,
    collection
  };
};


export default listingActions;