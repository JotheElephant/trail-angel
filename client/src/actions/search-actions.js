import * as _ from 'lodash';

import actionTypes from './action-types';
import dataApi from '../api';

const submitSearch = (term) => {
  return {
    type: actionTypes.SUBMIT_SEARCH,
    term
  };
};

const receiveSearchResults = (results) => {
  return {
    type: actionTypes.RECEIVE_SEARCH_RESULTS,
    results: results
  };
};

// mutates the trail to include a isFavorite flag
const associateFavorites = (trails, favorites) => {
  _.map(trails, (trail) => {
    trail.isFavorite = _.findIndex(favorites, { id: trail.id }) !== -1;
    trail.isSearchResult = true;
  });
};

export const cancelSearch = () => {
  return (dispatch) => {
    dispatch({
      type: actionTypes.CANCEL_SEARCH
    });
    dispatch(receiveSearchResults([]));
  };
}

export const search = (options) => {
  return (dispatch, getState) => {
    if (!getState().searchReducer.isCancelled) {
      dispatch(submitSearch(options));
      return dataApi.googlePlaces.search(options)
        .then((results) => {
          associateFavorites(results, getState().favoritesReducer.favorites);
          return dispatch(receiveSearchResults(results));
        })
        .catch((err) => {
          console.error('error retrieving search data', err);
        });
    }
  };
};