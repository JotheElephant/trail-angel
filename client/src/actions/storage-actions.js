'use strict';
import { AsyncStorage } from 'react-native';
import actionTypes from './action-types';
import dataApi from '../api';

const storageActions = {
  loadListingsFromStorage: () => {
    return (dispatch) => {
      AsyncStorage.getAllKeys()
        .then((keys) => {
          return AsyncStorage.multiGet(keys);
        })
        .then((stores) => {
          if (!stores) return;

          const listings = {};

          stores.forEach((store) => {
            const data = JSON.parse(store[1]);
            const keyName = store[0].replace(/^listing:/, '');

            if (/^listing:/.test(store[0])) {
              listings[keyName] = data;
            }
          });

          return listings;
        })
        .then((listings) => {
          dispatch({
            type: actionTypes.LOAD_LISTINGS_FROM_STORAGE,
            listings
          });
        })
        .catch((err) => {
          console.error('Error loading data from local storage: ', err);
        });
    };
  },

  saveToStorage: (listing) => {
    return (dispatch) => {
      const keyName = `listing:${listing.id}`;
      AsyncStorage.setItem(keyName, JSON.stringify(listing))
        .then(() => {
          dispatch ({
            type: actionTypes.SAVE_TO_STORAGE,
            data: listing
          });
          return listing;
        });
    };
  },

  removeFromStorage: (id) => {

  },

  clearAllListingsFromStorage: () => {

  }
};

export default storageActions;