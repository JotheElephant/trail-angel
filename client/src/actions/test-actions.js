import * as actionTypes from './action-types';
import * as dataApi from '../api';

const update = (custom, actionType) => {
  return {
    type: actionType,
    custom: custom
  };
}

export const testIncrement = (custom) => {
  // traditional action creator style
  return {
    type: actionTypes.TEST_INCREMENT,
    custom: custom
  };
};

export const testDecrement = (custom) => {
  // traditional action creator style
  return {
    type: actionTypes.TEST_DECREMENT,
    custom: custom
  };
};

export const testIncrementAsync = (custom) => {
  // by returning a function, the Thunk middleware is invoked
  return (dispatch, getState) => {
    // you could do something async here before dispatching
    // or only dispatch conditionally

    dataApi.yelp.searchYelp()
      .then((data) => {
        dispatch({
          type: 'TEST_INCREMENT',
          custom: JSON.stringify(data)
        })
      });
  };
}

export const getYelpData = () => {
  return {
    type: actionTypes.GET_YELP_DATA
  };
};