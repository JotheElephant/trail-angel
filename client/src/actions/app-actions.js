'use strict';

import actionTypes from './action-types';
import * as userActions from './user-actions';
import * as trailActions from './trail-actions';
import * as favoriteActions from './favorite-actions';

const appActions = {
  getGeolocation: (options = {
    enableHighAccuracy: true,
    timeout: 20000,
    maximumAge: 1000}
  ) => {

    return (dispatch) => {
      dispatch({
        type: actionTypes.GET_GEOLOCATION
      });

      return new Promise((resolve, reject) => {
        navigator.geolocation.getCurrentPosition(
          (position) => {
            resolve(dispatch(receiveGeolocation(position)));
          },
          (err) => {
            console.error('Error obtaining geolocation: ', err);
            reject(err);
          },
          options
        );
      });

    };
  },

  initializeApp: (profile) => {
    return (dispatch, getState) => {
      dispatch({
        type: actionTypes.INITIALIZE_APP
      });
      return dispatch(userActions.loginUser(profile))
        .then(() => {
          return dispatch(favoriteActions.fetchFavorites());
        })
        .then(() => {
          return dispatch(appActions.getGeolocation());
        })
        .then(() => {
          return dispatch(trailActions.fetchTrails({
            latitude: getState().appReducer.geolocation.coords.latitude,
            longitude: getState().appReducer.geolocation.coords.longitude
          }));
        })
        .catch((err) => {
          console.error('Error initializing application: ', err);
        });
    };
  }
};

const receiveGeolocation = (geolocation) => {
  return {
    type: actionTypes.RECEIVE_GEOLOCATION,
    geolocation
  };
};

export default appActions;