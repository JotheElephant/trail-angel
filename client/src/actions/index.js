import testActions from './test-actions';
import * as trailActions from './trail-actions';
import favoriteActions from './favorite-actions';
import * as userActions from './user-actions';
import * as searchActions from './search-actions';

const actions = {
  trailActions,
  testActions,
  favoriteActions,
  userActions,
  searchActions
};

export default actions;