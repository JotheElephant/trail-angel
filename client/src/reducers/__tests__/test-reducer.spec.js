import testReducer, { initialState } from '../test-reducer';
import { testIncrement, testDecrement } from '../../actions/test-actions';

describe('testReducer', () => {
  it('handles TEST_INCREMENT action', () => {
    expect(testReducer(initialState, testIncrement('custom thing'))).toMatchSnapshot();
  });

  it('handles TEST_DECREMENT action', () => {
    expect(testReducer(initialState, testDecrement('something else')))
      .toMatchSnapshot();
  });
});