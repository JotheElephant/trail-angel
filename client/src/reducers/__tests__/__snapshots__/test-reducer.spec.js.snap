exports[`testReducer handles TEST_DECREMENT action 1`] = `
Object {
  "bar": "hello",
  "custom": "something else",
  "foo": 21,
}
`;

exports[`testReducer handles TEST_INCREMENT action 1`] = `
Object {
  "bar": "helloo",
  "custom": "custom thing",
  "foo": 84,
}
`;
